### DISCUSSION TITLE
Is there any way to retrieve multiple documents by ID?

### DISCUSSION THREAD

**Discussion by guplem** [firebase/flutterfire#9690] [Q&A] [answered]
In my flutter/dart app, I have a list of IDs from document (of the same collection) that I want to retrieve from FireStore.

I am trying to do the following:
`await FirebaseFirestore.instance.collection('groups').doc([id1, id2, id3, id4]).get();`

But ".doc()" won't accept a list as a parameter for the document ID to look for.

I tried with the "where" method like this:
`await FirebaseFirestore.instance.collection('groups').where("id", whereIn: [id1, id2, id3, id4]).get();`

But it won't work since the "id" isn't a field in the documents (it is their name/identifier).

The only two solutions I can think about are:
- Doing one request per document I want to retrieve (not ideal at all)
- Adding a field "id" in the documents that store the same information as their name/identifier (I don't like duplicating data that can end mismatching)

The third one (ideal IMO) would be to enable .doc to accept lists/arrays, so you can search for a list of documents you know the identifier of.

Am I missing something? Does anybody know any other solution?

THANKS!

(I have asked the same question [here](https://stackoverflow.com/questions/73986338/is-there-any-way-to-retrieve-multiple-documents-by-id-using-firestore-flutterfi))

**Comment by guplem** [comment, marked as answer]
I found a solution [here](https://stackoverflow.com/questions/46721517/google-firestore-how-to-get-several-documents-by-multiple-ids-in-one-round-tri/62148517#62148517)!

So adapting it to my case it looks like this:
`await FirebaseFirestore.instance.collection('groups').where(FieldPath.documentId, whereIn: [id1, id2, id3, id4]).get();
`

Be aware that there is a limit of 10 to the number of values/IDs you can use in the search.

**Comment by vogdb** [comment]
If you need more than 10 ids.

First create a function to fetch by ids:
```dart
Future<List<Book>> _getBooksByIds(ids) {
  return _instance
      .collection('books')
      .where(FieldPath.documentId, whereIn: ids)
      .get()
      .then((QuerySnapshot snapshot) {
    return snapshot.docs.map((DocumentSnapshot doc) => BookModel.fromSnapshot(doc)).toList();
  });
}

```

Then split by 10 your ids
```dart
ids = ['id1', 'id2'...'id11'];
final idsBy10 = [
  for (var i = 0; i < ids.length; i += 10)
    ids.sublist(i, min(i + 10, ids.length))
];
```

Then run multiple requests and merge their results:
```dart
Future.wait<List<Book>>(idsBy10.map((ids) => _getBooksByIds(ids)))
        .then((listOfLists) => listOfLists.expand((l) => l).toList());
```

Please share If you have an alternative solution shorter than mine. I would be glad to use it.
